File: src/geom/Point2D.cpp
#include <cmath>

#include "../../include/geom/Point2D.h"
#include "../../include/utils/Tools.h"

Point2D::Point2D()
    : x{0.0}, y{0.0}
{}

Point2D::Point2D(double x, double y)
    : x{x}, y{y}
{}


Point2D::~Point2D()
{}

double Point2D::distanceSquaredTo(Point2D const &p) const
{
    double const DX = x - p.x;
    double const DY = y - p.y;

    return DX * DX + DY * DY;
}


double Point2D::distanceTo(Point2D const &p) const
{
    return std::sqrt(distanceSquaredTo(p));
}

bool Point2D::areColinear(Point2D const &p1, Point2D const &p2, Point2D const &p3)
{
    // We compute a fake area (we normally should divide by 2.0 but not needed here)
    return doubleAlmostEqualToZero(std::abs(p1.x * (p2.y - p3.y) + p2.x * (p3.y - p1.y) + p3.x * (p1.y - p2.y)));
}

std::ostream& operator<<(std::ostream &os, Point2D const &p)
{
    return os << "X: " << p.x << ", Y: " << p.y;
}

File: java/src/com/upmc/geom/Point2D.java
package com.upmc.geom;

public class Point2D {
	public double x;
	public double y;
	
	public Point2D(double x, double y) {
		this.x = x;
		this.y = y;
	}
	
	public double distanceSquaredTo(final Point2D p) {
		final double DX = x - p.x;
		final double DY = y - p.y;
		
		return DX * DX + DY * DY;
	}
	
	public double distanceTo(final Point2D p) {
		return Math.sqrt(distanceSquaredTo(p));
	}
	
	public static boolean areColinear(final Point2D p1, final Point2D p2, final Point2D p3) {
		return Math.abs(p1.x * (p2.y - p3.y) + p2.x * (p3.y - p1.y) + p3.x * (p1.y - p2.y)) == 0.0;
	}
	
	@Override
	public String toString() {
		return "X: " + x + ", Y: " + y; 
	}
}
